sapicore/engine/network/__init__.py

```diff
@@ -0,0 +1,400 @@
+                    self.graph.add_edge(*synapse.identifier.split("->"), reference=synapse)
+                else:
+                    logging.info(f"Could not add synapse, invalid path given: {synapse}")
+                    continue
+            else:
+                raise TypeError("Synapses must be given as object references, dictionaries, or paths to YAML.")
+
+            # # heterogenize modifies the object iff a `sweep` specification was provided in the model configuration.
+            synapse.heterogenize(num_combinations=synapse.num_units)
+
+        # update root node list.
+        self._find_roots()
+
+    def get_ensembles(self):
+        """Returns all ensemble references in the network `graph`."""
+        return [self.graph.nodes[i]["reference"] for i in self.graph.nodes]
+
+    def get_synapses(self):
+        """Returns all synapse references in the network `graph`."""
+        return [self.graph.edges[s, d]["reference"] for s, d in self.graph.edges]
+
+    def build(self):
+        """Adds and initializes named components to the graph from this instance's `configuration` dictionary.
+
+        Note
+        ----
+        Network YAMLs reference ensemble and synapse YAMLs by their path relative to that specified in the `root`
+        field (or the project root if that field was not provided).
+
+        """
+        # set network identifier string from configuration file.
+        self.identifier = self.configuration.get("identifier", "net")
+
+        # use network API to add ensembles and synapses based on configuration dictionary or YAML paths.
+        if "ensembles" in self.configuration.get("model"):
+            ensemble_paths = [
+                os.path.join(self.configuration.get("root", ""), ep)
+                for ep in self.configuration.get("model", {}).get("ensembles")
+            ]
+            self.add_ensembles(*ensemble_paths)
+
+        if "synapses" in self.configuration.get("model"):
+            synapse_paths = [
+                os.path.join(self.configuration.get("root", ""), sp)
+                for sp in self.configuration.get("model", {}).get("synapses")
+            ]
+            self.add_synapses(*synapse_paths)
+
+    def data_hook(self, data_dir: str, steps: int, *args: Component):
+        """Attach a data accumulator forward hook to some or all network components.
+
+        Parameters
+        ----------
+        data_dir: str
+            Path to directory in which to save intermediate simulation output.
+
+        steps: int
+            Total number of simulation steps in the experiment to be logged.
+            Required for HDF5 sizing and chunk management.
+
+        args: Component, optional
+            Components to attach data hooks to. If not provided, data will be logged for all components.
+
+        """
+        if not args:
+            for ensemble in self.get_ensembles():
+                DataAccumulatorHook(ensemble, data_dir, ensemble.get_loggable(), steps)
```

Comment:
You should probably accumulate and return these hook classes in a list. Otherwise, a) they may be autodeleted by python garbage collection randomly if there's no ref to it anywhere. And b) users may want them anyway to interact with the hooks.